examples/pure_pursuit_controller/pure_pursuit_controller.cc

```diff
@@ -16,187 +16,202 @@
-
-            // we store the physical unit coordinate in the waypoint list
-            display.pixelToMM(xp, yp,  xMM,  yMM);
-            wpCoordinates.emplace_back(xMM,yMM);
-        }
+public:
+    PPCExample()
+      : m_Robot(MaxSpeed, 500_mm, 0_m, MaxTurn)
+      , m_Controller(LookAheadDistance, m_Robot.getDistanceBetweenAxes(), StoppingDist)
+      , m_Display({ 10.2_m, 10.2_m })
+      , m_CarSprite(m_Display.createCarAgent(160_mm))
+      , m_WayPointLines(m_Display.createLineStrip(sf::Color::Blue))
+      , m_LookAheadLine(m_Display.createLineStrip(sf::Color::Red))
+      , m_StateMachine(this, Invalid)
+    {
+        // Add an initial waypoint in the centre of the window
+        addWayPoint(sf::Vector2i{ World::WindowWidth / 2, World::WindowHeight / 2 });
+    }
 
-        // set waypoints in controller
-        controller.setWayPoints(wpCoordinates);
+    void run()
+    {
+        // Initially control with keyboard
+        m_StateMachine.transition(ControllerOff);
 
-        // run a GUI step
-        const auto key = display.runGUI(car.getPose());
+        // Run until window is closed
+        while (m_Display.isOpen()) {
+            m_StateMachine.update();
+        }
+    }
 
-        // clear screen before drawing other elements
-        display.clearScreen();
+private:
+    using Event = FSM<State>::StateHandler::Event;
+
+    SimulatedAckermann m_Robot;                 // simulated robot
+    Robots::PurePursuitController m_Controller; // for auto-controlling the robot
+    World m_Display;                            // display the agent
+    World::CarAgent m_CarSprite;
+    World::LineStrip m_WayPointLines;
+    std::vector<sf::RectangleShape> m_WayPointRectangles;
+    World::LineStrip m_LookAheadLine;
+    FSM<State> m_StateMachine;
+
+    template<class T>
+    void addWayPoint(const T& position)
+    {
+        constexpr float SquareWidth = 6.f;
+
+        // add rectangle for new way point
+        sf::RectangleShape rect{ sf::Vector2f{ SquareWidth, SquareWidth } };
+        rect.setFillColor(sf::Color::Blue);
+        rect.setPosition(position.x - SquareWidth / 2.f, position.y - SquareWidth / 2.f);
+        m_WayPointRectangles.push_back(rect);
+
+        // add point to end of line
+        m_WayPointLines.append(sf::Vector2f{ (float) position.x, (float) position.y });
+
+        // tell controller about way point
+        m_Controller.addWayPoint(m_Display.pixelToVector(position.x, position.y));
+    }
 
-        // draw all the waypoints on the screen
-        for (auto &r : rekt_list) {
-            display.drawRectangleAtCoordinates(r);
+    virtual bool handleEvent(State state, Event event) override
+    {
+        if (state == Invalid) {
+            return true;
         }
 
-        // draw lines between waypoints to form a path
-        if (rekt_list.size() > 1) {
-            drawLinesBetweenRects(rekt_list, display.getRenderer());
-        }
+        switch (event) {
+        case Event::Enter:
```

Comment:
Ifs would probably make sense here - imo swith is only preferable when each case is a one liner otherwise it's easy to lose track of the stupid fall-through

Reply:
That's less of a problem with recent versions of gcc though. They give you a warning if you don't have ``// falls through`` or similar in there. So our CI will pick up these sorts of problems.